reconcile/unleash_feature_toggles/integration.py

```diff
@@ -0,0 +1,310 @@
+    """Raised when a feature toggle is manually created."""
+
+
+class UnleashFeatureToggleDeleteError(Exception):
+    """Raised when a feature toggle is not marked for deletion."""
+
+
+class UnleashTogglesIntegration(
+    QontractReconcileIntegration[UnleashTogglesIntegrationParams]
+):
+    """Manage Unleash feature toggles."""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_early_exit_desired_state(
+        self, query_func: Callable | None = None
+    ) -> dict[str, Any]:
+        """Return the desired state for early exit."""
+        if not query_func:
+            query_func = gql.get_api().query
+        return {
+            "toggles": [ft.dict() for ft in self.get_unleash_instances(query_func)],
+        }
+
+    def get_unleash_instances(
+        self, query_func: Callable, instance: str | None = None
+    ) -> list[UnleashInstanceV1]:
+        """Get all Unleash instances with their projects and feature toggles."""
+        instances = [
+            ui
+            for ui in unleash_instances_query(query_func).instances or []
+            if (not instance or ui.name == instance) and ui.admin_token
+        ]
+
+        # Set default values for all feature toggles
+        for inst in instances:
+            for project in inst.projects or []:
+                for feature_toggle in project.feature_toggles or []:
+                    feature_toggle.unleash.q_type = (
+                        feature_toggle.unleash.q_type or "release"
+                    )
+                    feature_toggle.unleash.impression_data = (
+                        feature_toggle.unleash.impression_data or False
+                    )
+        return instances
+
+    def fetch_current_state(self, client: UnleashServer) -> list[Project]:
+        """Fetch the current state of all Unleash projects including their feature toggles."""
+        return client.projects(include_feature_toggles=True)
+
+    def get_project_by_name(
+        self, projects: Iterable[UnleashProjectV1], name: str
+    ) -> UnleashProjectV1:
+        for project in projects:
+            if project.name.lower() == name.lower():
+                return project
+        raise ValueError(f"Project {name} not found")
+
+    def validate_unleash_projects(
+        self,
+        instance_name: str,
+        current_projects: Iterable[Project],
+        desired_projects: Iterable[UnleashProjectV1],
+    ) -> None:
+        """Validate that all projects referenced in the desired state are actually exist."""
+        current_project_names = {p.name.lower() for p in current_projects}
+        desired_project_names = {p.name.lower() for p in desired_projects}
+        if missing_projects := desired_project_names - current_project_names:
+            for p in missing_projects:
+                logging.error(f"[{instance_name}] Project '{p}' does not exist!")
+            raise ValueError(f"Non-existing projects: {missing_projects}")
```

Comment:
here we are actually validating 2 cases, desired project missing and current project missing, how about do both in one validation, or just diff them and use later like this:

```python
                diff = diff_any_iterables(
                    current=projects,
                    desired=instance.projects or [],
                    current_key=lambda c: c.name.lower(),
                    desired_key=lambda d: d.name.lower(),
                    equal=lambda c, d: c.name.lower() == d.name.lower(),
                )
                if diff.add:
                    raise ValueError(f"missing projects: {','.join(diff.add.keys())}")
                if diff.delete:
                    raise ValueError(
                        f"not found projects: {','.join(diff.delete.keys())}"
                    )
                for pair in diff.identical.values():
                    self.reconcile(
                        client=client,
                        instance=instance,
                        project_id=pair.current.pk,
                        dry_run=dry_run,
                        current_state=pair.current.feature_toggles,
                        desired_state=pair.desired.feature_toggles or [],
                    )
```

Reply:
Changed in the last commit.